File: src/domain/equipment.ts
/**
 * Gear.
 *
 * Equipment contributes flat stats, may carry an elemental alignment that
 * feeds the wearer's power table, and may proc an art on a normal attack.
 * Procs are what keep basic attacks from going dead in the late game.
 */

import type { Element } from './elements.js';
import type { StatModifier } from './stats.js';

export const EQUIP_SLOTS = ['weapon', 'armor', 'helm', 'shield', 'trinket'] as const;
export type EquipSlot = (typeof EQUIP_SLOTS)[number];

export interface GearProc {
  /** Art resolved when the proc fires. */
  artId: string;
  /** Probability per qualifying attack. */
  chance: number;
  /** Power multiplier applied to the art on top of its own. */
  powerScale?: number;
}

export interface GearDef {
  id: string;
  name: string;
  description?: string;
  slot: EquipSlot;
  modifier?: StatModifier;
  element?: Element;
  /** Fires on a normal attack. */
  proc?: GearProc;
  /** Actors allowed to equip it. Empty means anyone. */
  restrictedTo?: string[];
  value?: number;
}

export type Loadout = Partial<Record<EquipSlot, string>>;

export function equippedIds(loadout: Loadout): string[] {
  const ids: string[] = [];
  for (const slot of EQUIP_SLOTS) {
    const id = loadout[slot];
    if (id) ids.push(id);
  }
  return ids;
}

export function canEquip(gear: GearDef, actorId: string): boolean {
  if (!gear.restrictedTo || gear.restrictedTo.length === 0) return true;
  return gear.restrictedTo.includes(actorId);
}

export function gearModifiers(
  loadout: Loadout,
  defOf: (id: string) => GearDef | undefined,
): StatModifier[] {
  const mods: StatModifier[] = [];
  for (const id of equippedIds(loadout)) {
    const def = defOf(id);
    if (def?.modifier) mods.push(def.modifier);
  }
  return mods;
}

/** Procs available from currently worn gear, in slot order. */
export function gearProcs(
  loadout: Loadout,
  defOf: (id: string) => GearDef | undefined,
): GearProc[] {
  const procs: GearProc[] = [];
  for (const id of equippedIds(loadout)) {
    const def = defOf(id);
    if (def?.proc) procs.push(def.proc);
  }
  return procs;
}

File: unity/Aetherlight.Core/Domain/Equipment.cs
using System;
using System.Collections.Generic;

namespace Aetherlight.Domain
{
    public enum EquipSlot
    {
        Weapon,
        Armor,
        Helm,
        Shield,
        Trinket,
    }

    /// <summary>Fires on a normal attack, which keeps basic attacks alive late.</summary>
    public sealed class GearProc
    {
        public string ArtId = "";
        public double Chance;
        public double PowerScale = 1;
    }

    public sealed class GearDef
    {
        public string Id = "";
        public string Name = "";
        public string? Description;
        public EquipSlot Slot;
        public StatModifier? Modifier;
        public Element? Element;
        public GearProc? Proc;
        /// <summary>Actors allowed to equip it. Empty means anyone.</summary>
        public List<string> RestrictedTo = new List<string>();
        public double Value;
    }

    /// <summary>What an actor is wearing, by slot.</summary>
    public sealed class Loadout
    {
        private readonly Dictionary<EquipSlot, string> _slots = new Dictionary<EquipSlot, string>();

        public string? this[EquipSlot slot]
        {
            get => _slots.TryGetValue(slot, out var id) ? id : null;
            set
            {
                if (value == null) _slots.Remove(slot);
                else _slots[slot] = value;
            }
        }

        public IEnumerable<string> EquippedIds()
        {
            foreach (EquipSlot slot in Enum.GetValues(typeof(EquipSlot)))
            {
                if (_slots.TryGetValue(slot, out var id)) yield return id;
            }
        }

        public Loadout Clone()
        {
            var copy = new Loadout();
            foreach (var pair in _slots) copy[pair.Key] = pair.Value;
            return copy;
        }
    }

    public static class Equipment
    {
        public static bool CanEquip(GearDef gear, string actorId) =>
            gear.RestrictedTo.Count == 0 || gear.RestrictedTo.Contains(actorId);

        public static List<StatModifier> Modifiers(Loadout loadout, Func<string, GearDef?> defOf)
        {
            var mods = new List<StatModifier>();
            foreach (var id in loadout.EquippedIds())
            {
                var mod = defOf(id)?.Modifier;
                if (mod != null) mods.Add(mod);
            }
            return mods;
        }

        public static List<GearProc> Procs(Loadout loadout, Func<string, GearDef?> defOf)
        {
            var procs = new List<GearProc>();
            foreach (var id in loadout.EquippedIds())
            {
                var proc = defOf(id)?.Proc;
                if (proc != null) procs.Add(proc);
            }
            return procs;
        }
    }
}
